examples/state_machine/states.py

```diff
@@ -0,0 +1,471 @@
+        self.context = context
+        self.state = state
+        self.effect = effect
+
+
+class UnsubscribedState(PNState):
+    def __init__(self, context: PNContext) -> None:
+        super().__init__(context)
+        self._transitions = {
+            events.SubscriptionChangedEvent.__name__: self.subscription_changed,
+            events.SubscriptionRestoredEvent.__name__: self.subscription_restored,
+        }
+
+    def subscription_changed(self, event: events.SubscriptionChangedEvent, context: PNContext) -> PNTransition:
+        context.channels = event.channels
+        context.groups = event.groups
+        context.tt = 0
+        self._context.update(context)
+        return PNTransition(state=HandshakingState, context=self._context)
+
+    def subscription_restored(self, event: events.SubscriptionRestoredEvent, context: PNContext) -> PNTransition:
+        context.channels = event.channels
+        context.groups = event.groups
+        context.tt = event.timetoken
+        context.region = event.region
+        self._context.update(context)
+        return PNTransition(state=ReceivingState, context=self._context)
+
+
+class HandshakingState(PNState):
+    def __init__(self, context: PNContext) -> None:
+        super().__init__(context)
+        self._transitions = {
+            events.HandshakeFailureEvent.__name__: self.reconnecting,
+            events.DisconnectEvent.__name__: self.disconnect,
+            events.HandshakeSuccessEvent.__name__: self.handshaking_success,
+            events.SubscriptionRestoredEvent.__name__: self.subscription_restored,
+            events.SubscriptionChangedEvent.__name__: self.subscription_changed,
+        }
+
+    def on_enter(self, context: Union[None, PNContext]):
+        self._context.update(context)
+        super().on_enter(self._context)
+        return effects.HandshakeEffect(self._context.channels, self._context.groups)
+
+    def on_exit(self):
+        super().on_exit()
+        return effects.CancelHandshakeEffect()
+
+    def subscription_changed(self, event: events.SubscriptionChangedEvent, context: PNContext) -> PNTransition:
+        self._context.update(context)
+        self._context.channels = event.channels
+        self._context.groups = event.groups
+        self._context.tt = 0
+
+        if event.channels[0] == 'fail':
+            next_state = HandshakeReconnectingState \
+                if context.handshake_count < HandshakingState.MAX_HANDSHAKE_RETRY_COUNT \
+                else HandshakeFailedState
+        else:
+            next_state = HandshakingState
+
+        return PNTransition(
+            state=next_state,
+            context=self._context
+        )
+
+    def subscription_restored(self, event: events.SubscriptionRestoredEvent, context: PNContext) -> PNTransition:
+        self._context.update(context)
+        self._context.channels = event.channels
+        self._context.groups = event.groups
+        self._context.tt = 0
```

Comment:
You should take `tt` and `region` from the event. That's the difference between it and `SubscriptionChanged`